server/src/main/java/org/elasticsearch/index/mapper/ObjectMapper.java

```diff
@@ -176,26 +176,52 @@ public final void addDynamic(String name, String prefix, Mapper mapper, Document
             // we just add it as it is for sure a leaf mapper
             if (name.contains(".") == false || (subobjects.isPresent() && (subobjects.get() == Subobjects.DISABLED))) {
                 add(name, mapper);
-            } else {
-                // We strip off the first object path of the mapper name, load or create
-                // the relevant object mapper, and then recurse down into it, passing the remainder
-                // of the mapper name. So for a mapper 'foo.bar.baz', we locate 'foo' and then
-                // call addDynamic on it with the name 'bar.baz', and next call addDynamic on 'bar' with the name 'baz'.
-                int firstDotIndex = name.indexOf('.');
-                String immediateChild = name.substring(0, firstDotIndex);
-                String immediateChildFullName = prefix == null ? immediateChild : prefix + "." + immediateChild;
-                Builder parentBuilder = findObjectBuilder(immediateChildFullName, context);
-                if (parentBuilder != null) {
-                    parentBuilder.addDynamic(name.substring(firstDotIndex + 1), immediateChildFullName, mapper, context);
-                    add(parentBuilder);
-                } else if (subobjects.isPresent() && subobjects.get() == Subobjects.AUTO) {
-                    // No matching parent object was found, the mapper is added as a leaf - similar to subobjects false.
-                    add(name, mapper);
-                } else {
-                    // Expected to find a matching parent object but got null.
-                    throw new IllegalStateException("Missing intermediate object " + immediateChildFullName);
+                return;
+            }
+
+            if (subobjects.isPresent() && subobjects.get() == Subobjects.AUTO) {
+                // Check for parent objects. Due to auto-flattening, we need to check for all possible object names.
```

Comment:
I think it would help to understand this part of the code better if you make a simple example, similar to the one in the other code path.

Is this need because with subobjects: auto, we can't be sure whether the parent object is fully nested, dotted, or some mix?

Reply:
I agree, a comment would be nice.

Reply:
I am not sure i understand why is this different from `subobjects: false` logic. As far as i can tell actual flattening happens when mappers are built, meaning that when we work with builders here they have the same structure in both cases. What am i missing?

Reply:
I updated the comment and added an example, any better?

Reply:
I think i get it, it's because with `auto` we allow objects with dots in name to appear without intermediate objects being present.